### PR TITLE
if-then-else changes in Cmmgen

### PR DESCRIPTION
**Pull request by mshinwell**
I'm going to leave @chambart to comment on this.

### CODE REVIEW

**Comment by mshinwell** [comment]
(Marking as flambda-prerequisite until such time as @chambart may downgrade it.  It needs removing from chambart/flambda_merge if that happens.)

File: asmcomp/cmmgen.ml [hunk 1]
```diff
@@ -221,6 +221,13 @@ let untag_int = function
   | Cop(Cor, [c; Cconst_int 1]) -> Cop(Casr, [c; Cconst_int 1])
   | c -> Cop(Casr, [c; Cconst_int 1])
 
+let if_then_else (cond, ifso, ifnot) =
+  match cond with
+  | Cconst_int 0 -> ifnot
+  | Cconst_int 1 -> ifso
+  | _ ->
+    Cifthenelse(cond, ifso, ifnot)
```
Comment (chambart) [inline asmcomp/cmmgen.ml:229]: Note that the only case where this matters with flambda is with loops when the first rounds is always runned

**Comment by alainfrisch** [comment]
This looks correct to me. I'd still be interested in an example about the "for loop" unrolling requiring the simplification in if-then-else.  I would have expected this optimization to be done at the flambda level.

**Comment by chambart** [comment]
Hum sorry I looked at it a bit more and mixed something in my example.
In fact, the places where constant branch where not eliminated by flambda before, is when `exit_if_false` or `exit_if_true` is involved on code generated by transforming a sequand or sequor.

A more general transformation handling branches will be done at some point in flambda, but not in this release, and this small change allow to simply avoid some bad cases.

By the way, I'm still not completely convinced that the recursive cases in test_bool
(below let and if_then_else) are beneficial. That allows to improve some cases, but might
prevent the generation of efficient test sequence for the if (handled by selectgen), since the
test does not appear right below the if.

By the way, there is no loop (for, while) unrolling at all in flambda.

**Comment by chambart** [comment]
This is useful, but I wouldn't consider that a prerequisite, hence I removed the tag.

**Comment by damiendoligez** [comment]
OK for merging.

**Comment by alainfrisch** [comment]
@damiendoligez :  @chambart himself does not seem completely sure that this proposal cannot degrade some cases.  Shouldn't we apply some care before merging?

**Comment by mshinwell** [comment]
If we're worried about this, maybe what we should do is wait until a quiet time (maybe at a weekend) and then merge this.  OCamlPro's benchmark suite should be able to detect any noticeable regressions.

**Comment by damiendoligez** [comment]
OK for merging after getting some evidence that it doesn't degrade performance, or after removing the recursive cases.

Or, @chambart what do you think of postponing to 4.04?

**Comment by chambart** [comment]
It is not on my repository, so I can't change it, but I would be in favor of taking the obviously beneficial case and wait for 4.04 to have some time to test and think about the rest.

**Comment by mshinwell** [comment]
@chambart Just make a new pull request that reverts the bit you want to keep back...

**Comment by damiendoligez** [comment]
I'll re-approved when the changes are made.

**Comment by mshinwell** [comment]
@chambart Please confirm the diff now looks correct (I merged your PR on my repo and fixed a conflict).

@damiendoligez Please approve.

**Comment by mshinwell** [comment]
@chambart Just waiting for your approval too

**Comment by chambart** [comment]
This looks ok now.

**Pull request merged**